java/0004Sum.java

```diff
@@ -0,0 +1,228 @@
+            int sum = numbers[i] + numbers[j] + numbers[third] + numbers[fourth];
+            if (sum < target) {
+              third++;
+            } else if (sum > target) {
+              fourth--;
+            } else {//sum == target
+              ArrayList<Integer> list = new ArrayList<Integer>();
+              list.add(numbers[i]);
+              list.add(numbers[j]);
+              list.add(numbers[third]);
+              list.add(numbers[fourth]);
+              rst.add(list);
+              third++;
+              fourth--;
+              while (third < fourth && numbers[third] == numbers[third - 1]) {
+                third++;
+              }
+              while (third < fourth && numbers[fourth] == numbers[fourth + 1]){
+                fourth--;
+              }
+            }
+          }
+        }
+      }
+      return rst;
+    }
+}
+
+
+/*
+NOT Complete yet. Has a order issue in HashSet
+http://lifexplorer.me/leetcode-3sum-4sum-and-k-sum/
+Thoughts:
+Utilize 2Sum.
+
+Notes 2017: no need to create a new class Pair. We can just use list in HashSet<List>, which compairs the list element and eleminate duplicated list.
+*/
+
+public class Solution {
+    //Create class Pair for HashSet<Pair> to use
+    class Pair {
+      Integer x;
+      Integer y;
+
+      public Pair(int x, int y){
+        this.x = x;
+        this.y = y;
+      }
+
+      @Override
+      public int hashCode(){
+        return this.x.hashCode() + this.y.hashCode();
+      }
+
+      @Override
+      public boolean equals(Object d) {
+        if (!(d instanceof Pair)) {
+            return false;
+        }
+        Pair p = (Pair)d;
+        return (this.x == p.x) && (this.y == p.y);
+      }
+    }
+
+    public ArrayList<ArrayList<Integer>> fourSum(int[] numbers, int target) {
+      ArrayList<ArrayList<Integer>> rst = new ArrayList<ArrayList<Integer>>(); 
+      if (numbers == null || numbers.length < 4) {
+        return rst;
+      }
+      Arrays.sort(numbers);
+      HashMap<Integer, HashSet<Pair>> map = new HashMap<Integer, HashSet<Pair>>();
+      for (int i = 0; i < numbers.length; i++) {
+        for (int j = i + 1; j < numbers.length; j++) {
+          int sum = numbers[i] + numbers[j];
+          if (map.containsKey(target - sum)) {
+            for (Pair p : map.get(target - sum)) {
+              ArrayList<Integer> list = new  ArrayList<Integer>();
+              list.add(p.x);
+              list.add(p.y);
+              list.add(numbers[i]);
+              list.add(numbers[j]);
+              if (!rst.contains(list)) {
+                rst.add(list);
+              }
+            }
+          }
+        }
+        //Add all pairs up to i
+        for (int j = 0; j < i; j++) {
+          int sum = numbers[i] + numbers[j];
+          if (!map.containsKey(sum)) {
+            map.put(sum, new HashSet<Pair>());
+          }
+          map.get(sum).add(new Pair(numbers[j], numbers[i]));
+        }
+      }
+
+      return rst;
+    }

```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Syntax error, insert &#34;}&#34; to complete ClassBody